File: sol/solution/0001-0100/0025. Reverse Nodes in k-Group/Solution.php
/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution {
    /**
     * @param ListNode $head
     * @param Integer $k
     * @return ListNode
     */
    function reverseKGroup($head, $k) {
        $curr = $head;
        $count = 0;
        while ($curr !== null && $count < $k) {
            $curr = $curr->next;
            $count++;
        }
        if ($count === $k) {
            $reversedHead = $this->reverseLinkedList($head, $k);
            // Recursively process the remaining list
            $head->next = $this->reverseKGroup($curr, $k);
            return $reversedHead;
        }
        return $head;
    }
    /**
     * Reverses a linked list of size k and returns the new head.
     * @param ListNode $head
     * @param Integer $k
     * @return ListNode
     */
    private function reverseLinkedList($head, $k) {
        $prev = null;
        $curr = $head;
        while ($k > 0) {
            $next = $curr->next;
            $curr->next = $prev;
            $prev = $curr;
            $curr = $next;
            $k--;
        }
        return $prev;
    }
}

File: sol/solution/0001-0100/0025. Reverse Nodes in k-Group/Solution.scala
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
    def reverseKGroup(head: ListNode, k: Int): ListNode = {
        var curr: ListNode = head
        var count: Int = 0
        while (curr != null && count < k) {
            curr = curr.next
            count += 1
        }
        if (count == k) {
            val reversedHead: ListNode = reverseLinkedList(head, k)
            // Recursively process the remaining list
            head.next = reverseKGroup(curr, k)
            return reversedHead
        }
        head
    }
    def reverseLinkedList(head: ListNode, k: Int): ListNode = {
        var prev: ListNode = null
        var curr: ListNode = head
        var remainingK: Int = k
        while (remainingK > 0 && curr != null) {
            val next: ListNode = curr.next
            curr.next = prev
            prev = curr
            curr = next
            remainingK -= 1
        }
        prev
    }
}